src/core/drive/visit.js

```diff
@@ -219,7 +219,20 @@ export class Visit {
           this.adapter.visitRendered(this)
           this.complete()
         } else {
+          const currentSnapshot = this.view.snapshot
+          const preserveScroll = currentSnapshot.refreshScroll === "preserve"
+          const scrollPosition = preserveScroll ? { x: window.scrollX, y: window.scrollY } : null
+
           await this.view.renderError(PageSnapshot.fromHTMLString(responseHTML), this)
+
+          // Handle scroll position after error rendering
+          if (preserveScroll && scrollPosition) {
+            this.view.scrollToPosition(scrollPosition)
+          } else if (!this.scrollToAnchor()) {
+            this.view.scrollToTop()
+          }
+          this.scrolled = true
+
```

Comment:
Rather than adding these lines to account for the scrolling behavior, have you explored invoking `this.performScroll()`?

```suggestion
          await this.view.renderError(PageSnapshot.fromHTMLString(responseHTML), this)
          this.performScroll()
          
```

It would mirror this branch's counterpart (the `this.renderPageSnapshot(snapshot, false)`)

https://github.com/hotwired/turbo/blob/fea2ebd1ad9b92416b14e14c47da9e229c242ec9/src/core/drive/visit.js#L423-L428